tests/test_rsu_to_raw_ingest.py

```
@@ -0,0 +1,48 @@
+import sys
+sys.path.append("..")
+from GCP_cloud_functions.rsu_to_raw_ingest import rsu_to_raw_ingest
+import mock
+import base64
+import datetime
+from unittest.mock import patch, mock_open
+from unittest import TestCase
+from google.cloud.storage import Client
+from google.cloud.pubsub_v1 import PublisherClient
+import pytest
+
+"""
+This script handles the unit testing for the rsu-to-raw-ingest Cloud Function.
+"""
+
+mock_context = mock.Mock()
+mock_context.event_id = '617187464135194'
+mock_context.timestamp = '2019-07-15T22:09:03.761Z'
+
+@mock.patch("google.cloud.storage.Client")
+def test_PubSubmitted(client):
+    
+    name = 'Testing GCP automation, 1-2-3!'
+    data = {'data': base64.b64encode(name.encode())}
+    raw_bucket = client().get_bucket
+    rsu_to_raw_ingest.rsu_to_raw_ingest(data, mock_context)
+    blob = raw_bucket().blob
+    blob_name = (str(datetime.datetime.now().strftime("%Y-%m-%d %H:%M")))
```

This test method doesn't seem to be set up correctly. Several of the functions from above have been turned into properties and aren't working correctly.

We'll also want to raise an exception, not just test for one. Look into `side_effect` and how to use that to raise an exception when the `blob()` function is called.

---

Note for all exception-related testing: Spoke with Drew. He doesn't seem to find anything wrong with the syntax of my unit test for this exception, but he did say that if the tests are finishing before they throw an error then that would explain why you aren’t getting an error. Even though that exception has been thrown correctly in the past (such as when I accidentally forgot to use the right bucket name). So...I am still a little stumped here, but I'll keep going. There's another patch of code in the raw-to-data-lake.py (the "if is_json_true()" block) that I'll aim to cover to go above my current 89.2% coverage. 

He also told me that if I could get my coverage above 90% without the exceptions that I could just go without full coverage on the exception.

---

If we need to roll this unit testing task over to the next sprint we can do that as well. The coverage percentage is a little misleading here. We want to cover the case that an exception is thrown so we know the code handles this.

---

Hmm. I think I may have to do that. I perhaps carelessly thought I was finished since I hit that 90.2% on Friday - and why I wasn't as worried about covering the Exceptions that early in the sprint. Could we push this into the next sprint?

---

Go ahead and add a comment to your unit testing task saying that we did not finish in this sprint and are rolling to the next one to ensure unit tests are covering edge cases, then we can move it to the next sprint.

---

Sounds good. I also made some pushes to cover some of the missed lines but because Travis has halted, I don't think it's updated the coverage in SonarCloud (passed locally on my machine)? So, the coverage might be slightly higher than 89.8? Unsure how to proceed with this new development.

---

I've covered the Exception testing for the latter two Cloud Functions under Drew's guidance. For this, again, this function's updated script and tests will come next PR (from story branch CV-66 to dev).